### PR TITLE
Stake pools changes to DB

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#712 

# Overview

- [x] I have added separate SQL schema 
- [x] I have added DBPoolLayer
- [x] I have added impl of MVar version of DBPoolLayer
- [x] I have added impl of Sqlite version of DBPoolLayer

# Comments

Properties to check in the next PR:
1. `putPoolProduction` of sequence of  `(poolId, SlotIdN)` where `N` is incrementing in Sqlite gives the same result upon `readPoolProduction` as when MVar impl
2. `putPoolProduction` slots being in given epoch and `readPoolProduction (Just epoch)` with this epoch should give proper results
3. `readPoolProduction (Just epoch)` for any epoch always should always only give values from a given epoch
4. When we `putPoolProduction` pools and non-duplicating slots in increasing order then we should have all pool values and all slot values when `readPoolProduction Nothing`.
5. When we `putPoolProduction` pools and non-duplicating slots in increasing order then
for each pool id we have slots value that are diminishing in time
6. `rollbackTo slotIdN` should not have slot values later than slotIdN when `readPoolProduction Nothing)  
7. `rollbackTo slotIdN` should not affect slots not later than slotIdN
8. `readPoolProduction` should not return pools with no slots, with or without rollback

### CODE REVIEW

**Comment by KtorZ** [comment]
@paweljakubas We do now store each wallet in separate databases. So we probably need _another_ database for stake pools data. 

@Anviking what do you think? I haven't yet much thought into this but maybe you have already. I see two options at the moment:

1/ Have a separate database (and database interface / db layer) for stake pools. Such that, chain following and data storage is very much isolated from the wallets. Then, we can have a single worker doing that for all stake pools. 

2/ Extend the wallet layer and current db layer with stake pools tracking, such that each wallet worker does the tracking itself in its own database.

I am VERY MUCH suggesting the first approach, especially because we already had requests from other team to make this code easily "extractable" for the explorer or midnight. Plus, I don't see any reasons to "bloat" the wallet layer with stake pool stuff. Actually, we could even put everything regarding stake pools data tracking into a separate namespace `Cardano.StakePools.` and have a 
`Cardano.StakePools.DBLayer`.

**Comment by Anviking** [comment]
> what do you think? I haven't yet much thought into this but maybe you have already. I see two options at the moment:

It uncomfortably dawned upon me the other day, thanks for bringing it up. 😅 

In theory I *very* much like option 1, but I'm not sure if testing would make the transition to multiple types of DBs tricker.

I believe there's also a third option though:
3. Extend the existing DB, but have one extra shared database for it, containing just stake-pool metrics but no wallet stuff (yes, this is weird, but we could call it temporary)

**Comment by KtorZ** [comment]
>  3. Extend the existing DB, but have one extra shared database for it, containing just stake-pool metrics but no wallet stuff.

But then, both db would have unrelated tables & schemas coming from the other side :thinking: ? I do think keeping things separate will make for a better design (and easier testing!) here.

**Comment by Anviking** [comment]
> But then, both db would have the schema of the other DB 🤔 ? I do think keeping things separate will make for a better design (and easier testing!) here

Yes, in that case they would have the same schema.

I don't see how testing would be more difficult to do in option 3. I thought we be pretty good off with extending the state machine tests. And that there wasn't anything complicated to do… 🤔 Or hm, ok maybe it could be in some cases. It would be very bad if we began treating the DB in two different cases, un-enforced by the type-system. 🤔 

But I'm not sure how easy it would be to duplicate/generalize/rework the testing infrastructure as necessary for another DB. I imagine both of you would have a better feeling of that than me though.

A slight concern also:
If we start splitting up our databases, won't we eventually need guarantees that they work correctly together? Or maybe in this case it's nothing to worry about since wallets and StakePools never will interact closely together.

**Comment by KtorZ** [comment]
Testing and maintenance becomes easier because you have less areas of overlap. If we mix these two logics, then changing something in one might have a negative impact on the other.

Regarding existing testing, I think it's pretty straightforward to copy and adapt for the new database. There's no need to generalize anything, it's already pretty much generalized. Plus, properties that holds on one database might not be true for the other. Perhaps it would help to get a clear idea of what that database interface will be. I don't see much more than a put and read operation. So in that sense, the state-machine is maybe overkill here and we could get away with a few simple property based. 

--- 

>  If we start splitting up our databases, won't we eventually need guarantees that they work correctly together? Or maybe in this case it's nothing to worry about since wallets and StakePools never will interact closely together.

Hmm, I don't follow you on that? We'll be tracking two very different things that have no reason to be even particularly in sync. Ideally, we would want the stake pools DB to always be in front of any wallet databases but that's pretty much our only "consistency" concern here. The worse that can happen is an API call to fail with "Data not yet available, please retry in a bit".

**Comment by paweljakubas** [comment]
I am definitely in favor of first option : `1/ Have a separate database (and database interface / db layer) for stake pools.` We have a separate db for each wallet and one db for stake pools.

**Comment by KtorZ** [comment]
@paweljakubas: I am wondering about one thing:

- Should we actually consider **not** accepting a `Maybe EpochNumber` for the `readPoolProduction`, but only a `EpochNumber`. I don't think there's any use case at the moment where we would want to return all the stake pools, so we might as well no bother with it at all (smaller API, less things that can go wrong, less things to test). 

Do you agree?

File: lib/core/src/Cardano/DB/Sqlite.hs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+module Cardano.DB.Sqlite
+    ( SqliteContext (..)
+    , chunkSize
+    , dbChunked
+    , destroyDBLayer
+    , handleConstraint
+    , queryLogFunc
+    , startSqliteBackend
+    , transformTrace
+    , unsafeRunQuery
```
Comment (KtorZ) [inline lib/core/src/Cardano/DB/Sqlite.hs:31]: Not sure that we need to export everything here :thinking: ? Maybe it's best if some stuff remains hidden inside the module. I'd say: remove anything that isn't required by a dependent module :no_mouth:

File: lib/core/src/Cardano/DB/Sqlite.hs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    => Int -- ^ Chunk size
+    -> ([a] -> m b) -- ^ Action to run on values
+    -> [a] -- ^ The values
+    -> m ()
+chunkedM n f = mapM_ f . chunksOf n
+
+-- | Size of chunks when inserting, updating or deleting many rows at once. We
+-- only act on `chunkSize` values at a time. See also 'dbChunked'.
+chunkSize :: Int
+chunkSize = 100
```
Comment (KtorZ) [inline lib/core/src/Cardano/DB/Sqlite.hs:233]: It's actually nice to have all of that glue code all in the same place :)

File: lib/core/src/Cardano/Wallet/DB/Model.hs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE AllowAmbiguousTypes #-}
+1{-# LANGUAGE AllowAmbiguousTypes #-}
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DB/Model.hs]: whoops

**Review by KtorZ — COMMENTED** [review]
Looks good so far :+1:

File: lib/core/src/Cardano/Wallet/DB/Model.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/DB/Model.hs]: fixed

File: lib/core/src/Cardano/DB/Sqlite.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/DB/Sqlite.hs:31]: all is needed in other modules except `queryLogFunc` - fixed

**Review by KtorZ — APPROVED** [review]

**Comment by KtorZ** [comment]
bors retry

**Comment by rvl** [comment]
bors r+

#814 is merged now, so that CI error should never happen, regardless of which buildkite agent is used.

**Pull request merged**